File: java-code/src/main/java/refactoring/refactoring_api/replace_parameter_with_query/before/Item.java
package refactoring.refactoring_api.replace_parameter_with_query.before;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Getter
@AllArgsConstructor
public class Item {

  int quantity;
  double itemPrice;

  public static void main(String[] args) {
    Item anItem = new Item(120, 10.0);
    double finalPrice = anItem.finalPrice();
    log.info("Final Price: {}", finalPrice);
  }

  double finalPrice() {
    double basePrice = quantity * itemPrice;
    int discountLevel = (quantity > 100) ? 2 : 1;
    return discountedPrice(basePrice, discountLevel);
  }

  double discountedPrice(double basePrice, int discountLevel) {
    switch (discountLevel) {
      case 1:
        return basePrice * 0.95;
      case 2:
        return basePrice * 0.9;
      default:
        return basePrice;
    }
  }
}

File: go-code/refactoring_api/replace_parameter_with_query/before/item.go
package before

type Item struct {
	Quantity  int
	ItemPrice float64
}

func (item Item) finalPrice() float64 {
	basePrice := float64(item.Quantity) * item.ItemPrice
	var discountLevel int
	if item.Quantity > 100 {
		discountLevel = 2
	} else {
		discountLevel = 1
	}
	return item.discountedPrice(basePrice, discountLevel)
}

func (item Item) discountedPrice(basePrice float64, discountLevel int) float64 {
	switch discountLevel {
	case 1:
		return basePrice * 0.95
	case 2:
		return basePrice * 0.9
	default:
		return basePrice
	}
}

func main() {
	anItem := Item{Quantity: 120, ItemPrice: 10.0}
	finalPrice := anItem.finalPrice()
	println("Final Price:", finalPrice)
}
